devlib/collector/perfetto.py

```diff
@@ -0,0 +1,75 @@
+#    Copyright 2023 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import time
+import subprocess
+from shlex import quote
+
+from devlib.collector import (CollectorBase, CollectorOutput,
+                              CollectorOutputEntry)
+from devlib.exception import TargetStableError, HostError
+from devlib.utils.misc import check_output, which, memoized
+
+OUTPUT_PERFETTO_TRACE = 'trace.perfetto-trace'
+
+class PerfettoCollector(CollectorBase):
+
+    def __init__(self, target, config=None):
+        super().__init__(target)
+        self.target_output_file = target.path.join(self.target.working_directory, OUTPUT_PERFETTO_TRACE)
+        self.bg_cmd = None
+        self.config = config
+
+        if not self.target.is_installed('perfetto'):
+            raise TargetStableError('No perfetto found on device.')
+        self.target_binary = 'perfetto'
+
+    def start(self):
+        cmd = "cat {} | {} --txt -c - -o {}".format(
+            quote(self.config), quote(self.target_binary), quote(self.target_output_file)
+        )
+        self.start_time = time.time()
```

Comment:
We need to figure out why the pull times out as it's not really expected to happen instead of this hack. Also time.time() is not monotonic so `end - start < 0` is possible

Reply:
It's not even that it times out, more that there's an issue if we don't specify an explicit timeout?
```
> /home/kajpuc01/tools/lisa/external/devlib/devlib/collector/perfetto.py(69)get_data()
-> self.target.pull(self.target_output_file, self.output_path)
(Pdb) s
--Call--
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(202)__get__()
-> def __get__(self, *args, **kwargs):
(Pdb) n
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(203)__get__()
-> return self.__class__(
(Pdb)
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(204)__get__()
-> asyn=self.asyn.__get__(*args, **kwargs),
(Pdb)
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(205)__get__()
-> blocking=self.blocking.__get__(*args, **kwargs),
(Pdb)
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(203)__get__()
-> return self.__class__(
(Pdb)
--Return--
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(203)__get__()-><devlib.utils...x7f326301bc10>
-> return self.__class__(
(Pdb)
--Call--
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(208)__call__()
-> def __call__(self, *args, **kwargs):
(Pdb)
> /home/kajpuc01/tools/lisa/external/devlib/devlib/utils/asyn.py(209)__call__()
-> return self.blocking(*args, **kwargs)
(Pdb)
WARNING          Cancelling file transfer ['/sdcard/devlib-target/trace.perfetto-trace'] -> perfetto_test/wk1-speedometer-1/trace.perfetto-trace due to 'transfer inactive'
WARNING          Cancelling file transfer ['/sdcard/devlib-target/trace.perfetto-trace'] -> perfetto_test/wk1-speedometer-1/trace.perfetto-trace due to 'exception during transfer'
TimeoutError: adb -s 18251FDF60083A pull /sdcard/devlib-target/trace.perfetto-trace perfetto_test/wk1-speedometer-1/trace.perfetto-trace
```

I think the issue might be here:
```
    def _push_pull(self, action, sources, dest, timeout):
        # [...]
        if timeout or not self.poll_transfers:
            adb_command(self.device, command, timeout=timeout, adb_server=self.adb_server)
        else:
            with self.transfer_mgr.manage(sources, dest, action):
                bg_cmd = adb_command_background(self.device, command, adb_server=self.adb_server)
                self.transfer_mgr.set_transfer_and_wait(bg_cmd)
```

Depending on whether a timeout was specified we either use `adb_command` or `adb_command_background`. The former seems to work fine, the latter not so much in this case.